File: ttv/ttv.py
#!/usr/bin/python
# Same as ttv.sh but in Python
# Use through ttv wrapper

import os
import requests
import threading

URL = "https://www.twitch.tv/"
DEFAULT_RC_PATH = "~/.config/ttv/ttv.rc"


def parse_config():
    TTV_RC_PATH = os.path.expanduser(DEFAULT_RC_PATH)

    if os.path.exists(TTV_RC_PATH):
        with open(TTV_RC_PATH, "r") as TTV_RC:
            channel_names = TTV_RC.readlines()
        if len(channel_names) == 0:
            print("Config file empty")
            exit()
        return channel_names
    else:
        print(f"Config file not found ({TTV_RC_PATH})")
        exit()


def is_live(name):
    page = requests.get(URL + name).content.decode("utf-8")
    if "isLiveBroadcast" in page:
        print(name)


def main():
    threads = []
    channels = parse_config()

    for channel in channels:
        # name[:-1] to strip "\n" in link
        thread = threading.Thread(target=is_live, args=(channel[:-1],))
        thread.start()
        threads.append(thread)

    for thread in threads:
        thread.join()


main()

File: ttv/ttv.go
// Same as ttv.sh but in Go
// Use through ttv wrapper
package main

import (
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
)

func readConfig(path string) []string {
	f, err := os.ReadFile(path)
	if err != nil {
		log.Fatal(err)
	}
	return strings.Split(string(f), "\n")
}

func main() {
	const (
		TWITCH_URL       = "https://www.twitch.tv/"
		DEFAULT_CFG_PATH = "$HOME/.config/ttv/ttv.rc"
	)
	channels := readConfig(os.ExpandEnv(DEFAULT_CFG_PATH))

	var wg sync.WaitGroup
	wg.Add(len(channels))

	for _, channel := range channels {
		go func(channel string) {
			resp, err := http.Get(TWITCH_URL + channel)
			defer wg.Done()
			defer resp.Body.Close()
			if err != nil {
				log.Println(err)
			}
			page, err := io.ReadAll(resp.Body)
			if err != nil {
				log.Println(err)
			}
			if strings.Contains(string(page), "isLiveBroadcast") {
				fmt.Println(channel)
			}
		}(channel)
	}
	wg.Wait()
}
